# Fase 3 - Avance Semana 2  
## Notebook con scripts y mediciones de complejidad

Este notebook corresponde al avance formativo de la Fase 3 del proyecto transversal. Su objetivo es reutilizar el dataset procesado en la Fase 2 para implementar algoritmos estructurados y recursivos en Python, realizar mediciones básicas de complejidad temporal y documentar decisiones técnicas relacionadas con eficiencia, modularidad y escalabilidad.

El proyecto analiza el impacto de la inteligencia artificial en el empleo, la productividad y la capacitación laboral, utilizando variables como adopción de IA, riesgo de automatización, experiencia laboral, salarios antes y después de la IA, productividad y requerimientos de capacitación.

Este avance se vincula con la discusión desarrollada en el foro técnico de la Semana 1, donde se analizó la conveniencia de usar programación estructurada, recursiva o vectorizada según el tipo de problema, el volumen de datos y los criterios de eficiencia computacional.

In [1]:
from pathlib import Path
import sys
import timeit
import tracemalloc

import pandas as pd
import numpy as np

In [2]:
PROJECT_ROOT = Path.cwd().parents[1]

DATA_PROCESSED = PROJECT_ROOT / "data" / "processed" / "ai_job_impact_clean.csv"
F3_SRC = PROJECT_ROOT / "F3" / "src"
F3_RESULTS = PROJECT_ROOT / "F3" / "results"
RESULTS_COMPLEXITY = F3_RESULTS / "mediciones_complejidad_f3.txt"

sys.path.append(str(F3_SRC))

print("Ruta del proyecto:", PROJECT_ROOT)
print("Dataset procesado:", DATA_PROCESSED)
print("Carpeta scripts F3:", F3_SRC)
print("Archivo de resultados:", RESULTS_COMPLEXITY)

Ruta del proyecto: C:\Users\LuiskarEspinoza\Documents\Proyectos\mi_app\mcdi500_s1_grupo6
Dataset procesado: C:\Users\LuiskarEspinoza\Documents\Proyectos\mi_app\mcdi500_s1_grupo6\data\processed\ai_job_impact_clean.csv
Carpeta scripts F3: C:\Users\LuiskarEspinoza\Documents\Proyectos\mi_app\mcdi500_s1_grupo6\F3\src
Archivo de resultados: C:\Users\LuiskarEspinoza\Documents\Proyectos\mi_app\mcdi500_s1_grupo6\F3\results\mediciones_complejidad_f3.txt


In [3]:
df = pd.read_csv(DATA_PROCESSED)

print("Dataset cargado correctamente.")
print("Filas:", df.shape[0])
print("Columnas:", df.shape[1])

df.head()

Dataset cargado correctamente.
Filas: 2000
Columnas: 17


,employee_id,age,gender,education_level,industry,job_role,years_experience,ai_adoption_level,automation_risk,upskilling_required,salary_before_ai,salary_after_ai,job_status,work_hours_per_week,remote_work,job_satisfaction,productivity_change_%
0,E0001,50,Female,Bachelor,Marketing,Content Creator,26,High,High,Yes,106820,95455,Replaced,45,No,5,-10.64
1,E0002,45,Male,High School,Manufacturing,Quality Inspector,19,Low,Low,Yes,74131,72013,Unchanged,36,Yes,6,19.05
2,E0003,51,Female,Master,IT,DevOps Engineer,28,Medium,Medium,Yes,35311,42290,Modified,46,Yes,3,17.05
3,E0004,48,Male,PhD,Education,Teacher,24,Medium,Medium,Yes,114478,107820,Modified,50,No,9,-2.47
4,E0005,24,Male,Bachelor,Healthcare,Doctor,0,High,Medium,No,33890,40945,Modified,52,Yes,6,7.03


In [4]:
df.columns.tolist()

['employee_id',
 'age',
 'gender',
 'education_level',
 'industry',
 'job_role',
 'years_experience',
 'ai_adoption_level',
 'automation_risk',
 'upskilling_required',
 'salary_before_ai',
 'salary_after_ai',
 'job_status',
 'work_hours_per_week',
 'remote_work',
 'job_satisfaction',
 'productivity_change_%']

In [5]:
# Selección de variables relevantes para el análisis algorítmico de Fase 3

columnas_algoritmicas = [
    "age",
    "years_experience",
    "salary_before_ai",
    "salary_after_ai",
    "work_hours_per_week",
    "job_satisfaction",
    "productivity_change_%",
    "automation_risk",
    "ai_adoption_level",
    "upskilling_required"
]

df_algoritmico = df[columnas_algoritmicas].copy()

print("Dataset algorítmico creado correctamente.")
print("Filas:", df_algoritmico.shape[0])
print("Columnas:", df_algoritmico.shape[1])

df_algoritmico.head()

Dataset algorítmico creado correctamente.
Filas: 2000
Columnas: 10


,age,years_experience,salary_before_ai,salary_after_ai,work_hours_per_week,job_satisfaction,productivity_change_%,automation_risk,ai_adoption_level,upskilling_required
0,50,26,106820,95455,45,5,-10.64,High,High,Yes
1,45,19,74131,72013,36,6,19.05,Low,Low,Yes
2,51,28,35311,42290,46,3,17.05,Medium,Medium,Yes
3,48,24,114478,107820,50,9,-2.47,Medium,Medium,Yes
4,24,0,33890,40945,52,6,7.03,Medium,High,No


In [6]:
print("Valores nulos por columna:")
print(df_algoritmico.isnull().sum())

print("\nDuplicados:")
print(df_algoritmico.duplicated().sum())

print("\nTipos de datos:")
print(df_algoritmico.dtypes)

Valores nulos por columna:
age                      0
years_experience         0
salary_before_ai         0
salary_after_ai          0
work_hours_per_week      0
job_satisfaction         0
productivity_change_%    0
automation_risk          0
ai_adoption_level        0
upskilling_required      0
dtype: int64

Duplicados:
0

Tipos de datos:
age                        int64
years_experience           int64
salary_before_ai           int64
salary_after_ai            int64
work_hours_per_week        int64
job_satisfaction           int64
productivity_change_%    float64
automation_risk              str
ai_adoption_level            str
upskilling_required          str
dtype: object


## Preprocesamiento mínimo para análisis algorítmico

Para la Fase 3 se reutiliza el dataset procesado generado en la Fase 2. No se repite el pipeline completo de limpieza, ya que el objetivo de esta etapa es diseñar y medir algoritmos sobre datos previamente preparados.

Se seleccionaron variables numéricas y categóricas relacionadas con edad, experiencia laboral, salarios, productividad, riesgo de automatización, adopción de inteligencia artificial y requerimiento de capacitación. Estas variables permiten aplicar funciones estructuradas, algoritmos de ordenamiento, búsquedas y mediciones de complejidad temporal.

La validación inicial permite verificar que el subconjunto de datos utilizado para los algoritmos mantiene una estructura coherente y puede ser utilizado en las siguientes pruebas.

In [7]:
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

In [8]:
numeros = [29, 10, 14, 37, 13]
resultado = insertion_sort(numeros)
print("Lista ordenada:", resultado)

Lista ordenada: [10, 13, 14, 29, 37]


In [9]:
# Medición de tiempo del algoritmo insertion_sort

datos_prueba = df["salary_after_ai"].head(500).tolist()

tiempo_insertion = timeit.timeit(
    lambda: insertion_sort(datos_prueba.copy()),
    number=10
)

print("Tiempo de ejecución insertion_sort:", tiempo_insertion)
print("Cantidad de datos evaluados:", len(datos_prueba))

Tiempo de ejecución insertion_sort: 0.04684350000025006
Cantidad de datos evaluados: 500


In [10]:
df["salary_after_ai"].head(500).tolist()

[95455,
 72013,
 42290,
 107820,
 40945,
 104781,
 86183,
 105776,
 104197,
 42699,
 92231,
 109956,
 85267,
 64752,
 102088,
 36482,
 46193,
 42537,
 79219,
 101637,
 98204,
 35565,
 80986,
 47699,
 30347,
 82247,
 45216,
 141487,
 39587,
 30569,
 31875,
 47736,
 65237,
 40230,
 86765,
 76652,
 67480,
 107339,
 139250,
 73122,
 144902,
 56661,
 126241,
 85439,
 113428,
 70352,
 51741,
 98223,
 81832,
 82341,
 117348,
 90958,
 49707,
 85709,
 48429,
 91305,
 40515,
 64120,
 91794,
 97315,
 75487,
 85106,
 36437,
 97559,
 48329,
 47107,
 67973,
 115730,
 76957,
 46571,
 88278,
 124242,
 123049,
 41730,
 61727,
 99660,
 59726,
 35271,
 85089,
 52586,
 131012,
 78824,
 99694,
 87014,
 37588,
 92283,
 63588,
 67501,
 104135,
 45928,
 108468,
 54219,
 135554,
 67075,
 78770,
 89903,
 61751,
 82573,
 89278,
 89297,
 75089,
 113366,
 34030,
 77261,
 117250,
 43332,
 57048,
 107667,
 59325,
 38854,
 133082,
 78977,
 38043,
 45562,
 88030,
 79990,
 101032,
 148677,
 152101,
 118630,
 90369,
 28

In [13]:
# Medición de tiempo del algoritmo insertion_sort con datos reales del dataset

datos_prueba = df["salary_after_ai"].head(500).tolist()

tiempo_insertion = timeit.timeit(
    lambda: insertion_sort(datos_prueba.copy()),
    number=10
)

print("Tiempo de ejecución insertion_sort:", tiempo_insertion)
print("Cantidad de datos evaluados:", len(datos_prueba))

Tiempo de ejecución insertion_sort: 0.04388639999888255
Cantidad de datos evaluados: 500


In [14]:
insertion_sort(datos_prueba.copy())

[28345,
 28489,
 29117,
 30347,
 30569,
 30680,
 30937,
 31071,
 31576,
 31835,
 31875,
 33956,
 34030,
 34757,
 34825,
 35028,
 35239,
 35260,
 35271,
 35565,
 36437,
 36482,
 37210,
 37588,
 37692,
 37891,
 37938,
 38043,
 38195,
 38367,
 38854,
 39005,
 39150,
 39587,
 40212,
 40230,
 40413,
 40505,
 40515,
 40570,
 40748,
 40896,
 40945,
 40992,
 41041,
 41258,
 41292,
 41730,
 41883,
 42088,
 42290,
 42371,
 42429,
 42537,
 42567,
 42699,
 42752,
 42840,
 42936,
 42982,
 43142,
 43332,
 43362,
 43639,
 43861,
 44532,
 44890,
 45137,
 45216,
 45357,
 45562,
 45733,
 45791,
 45928,
 46188,
 46193,
 46571,
 46571,
 46680,
 47107,
 47691,
 47699,
 47736,
 47812,
 48163,
 48329,
 48429,
 48870,
 49229,
 49505,
 49506,
 49538,
 49601,
 49638,
 49673,
 49707,
 49826,
 49936,
 49963,
 50800,
 50821,
 51139,
 51273,
 51741,
 51795,
 52111,
 52541,
 52586,
 52712,
 52907,
 53022,
 53243,
 53405,
 53572,
 53778,
 53849,
 53930,
 53976,
 54023,
 54219,
 54764,
 54977,
 55159,
 55359,
 55504,


In [16]:
# Prueba simple del algoritmo con una lista pequeña

numeros_prueba = [29, 10, 14, 37, 13]

resultado_prueba = insertion_sort(numeros_prueba.copy())

print("Lista original:", numeros_prueba)
print("Lista ordenada:", resultado_prueba)

Lista original: [29, 10, 14, 37, 13]
Lista ordenada: [10, 13, 14, 29, 37]


In [17]:
# Medición de tiempo del algoritmo insertion_sort usando datos del dataset

datos_prueba = df["salary_after_ai"].head(500).tolist()

tiempo_insertion = timeit.timeit(
    lambda: insertion_sort(datos_prueba.copy()),
    number=10
)

print("Tiempo de ejecución insertion_sort:", tiempo_insertion)
print("Cantidad de datos evaluados:", len(datos_prueba))

Tiempo de ejecución insertion_sort: 0.04093200000170327
Cantidad de datos evaluados: 500


In [18]:
insertion_sort([29, 10, 14, 37, 13])

[10, 13, 14, 29, 37]

In [19]:
insertion_sort(datos_prueba.copy())

[28345,
 28489,
 29117,
 30347,
 30569,
 30680,
 30937,
 31071,
 31576,
 31835,
 31875,
 33956,
 34030,
 34757,
 34825,
 35028,
 35239,
 35260,
 35271,
 35565,
 36437,
 36482,
 37210,
 37588,
 37692,
 37891,
 37938,
 38043,
 38195,
 38367,
 38854,
 39005,
 39150,
 39587,
 40212,
 40230,
 40413,
 40505,
 40515,
 40570,
 40748,
 40896,
 40945,
 40992,
 41041,
 41258,
 41292,
 41730,
 41883,
 42088,
 42290,
 42371,
 42429,
 42537,
 42567,
 42699,
 42752,
 42840,
 42936,
 42982,
 43142,
 43332,
 43362,
 43639,
 43861,
 44532,
 44890,
 45137,
 45216,
 45357,
 45562,
 45733,
 45791,
 45928,
 46188,
 46193,
 46571,
 46571,
 46680,
 47107,
 47691,
 47699,
 47736,
 47812,
 48163,
 48329,
 48429,
 48870,
 49229,
 49505,
 49506,
 49538,
 49601,
 49638,
 49673,
 49707,
 49826,
 49936,
 49963,
 50800,
 50821,
 51139,
 51273,
 51741,
 51795,
 52111,
 52541,
 52586,
 52712,
 52907,
 53022,
 53243,
 53405,
 53572,
 53778,
 53849,
 53930,
 53976,
 54023,
 54219,
 54764,
 54977,
 55159,
 55359,
 55504,


In [20]:
# Prueba del algoritmo con datos reales del dataset

datos_salario = df["salary_after_ai"].head(20).tolist()

resultado_ordenado = insertion_sort(datos_salario.copy())

print("Primeros 20 salarios originales:")
print(datos_salario)

print("\nPrimeros 20 salarios ordenados:")
print(resultado_ordenado)

Primeros 20 salarios originales:
[95455, 72013, 42290, 107820, 40945, 104781, 86183, 105776, 104197, 42699, 92231, 109956, 85267, 64752, 102088, 36482, 46193, 42537, 79219, 101637]

Primeros 20 salarios ordenados:
[36482, 40945, 42290, 42537, 42699, 46193, 64752, 72013, 79219, 85267, 86183, 92231, 95455, 101637, 102088, 104197, 104781, 105776, 107820, 109956]


In [23]:
# Medición de tiempo del algoritmo insertion_sort usando datos reales del dataset

datos_prueba = df["salary_after_ai"].head(500).tolist()

tiempo_insertion = timeit.timeit(
    lambda: insertion_sort(datos_prueba.copy()),
    number=10
)

print("Tiempo de ejecución insertion_sort:", tiempo_insertion)
print("Cantidad de datos evaluados:", len(datos_prueba))

Tiempo de ejecución insertion_sort: 0.0513978999988467
Cantidad de datos evaluados: 500


In [24]:
# Comparación con sorted(), función optimizada de Python

tiempo_sorted = timeit.timeit(
    lambda: sorted(datos_prueba),
    number=10
)

print("Tiempo de ejecución insertion_sort:", tiempo_insertion)
print("Tiempo de ejecución sorted():", tiempo_sorted)

if tiempo_sorted > 0:
    print("Diferencia aproximada:", tiempo_insertion / tiempo_sorted, "veces")

Tiempo de ejecución insertion_sort: 0.0513978999988467
Tiempo de ejecución sorted(): 0.0002791999995679362
Diferencia aproximada: 184.08989999421664 veces


## Interpretación de la medición de complejidad

En esta prueba se comparó el algoritmo `insertion_sort`, implementado manualmente, con la función `sorted()` de Python, utilizando 500 registros reales de la variable `salary_after_ai` del dataset procesado en la Fase 2.

El resultado mostró que `insertion_sort` tuvo un tiempo de ejecución aproximado de 0.0514 segundos, mientras que `sorted()` obtuvo un tiempo aproximado de 0.00028 segundos. Esto indica que `sorted()` fue alrededor de 184 veces más rápido en esta prueba.

La diferencia se explica porque `insertion_sort` es un algoritmo simple, pero poco eficiente para conjuntos de datos grandes, ya que en el peor caso puede tener una complejidad temporal O(n²). Esto significa que, a medida que aumenta la cantidad de datos, el número de comparaciones puede crecer de forma considerable.

En cambio, `sorted()` utiliza una implementación optimizada de Python basada en Timsort, diseñada para ordenar datos de manera mucho más eficiente. Esta comparación evidencia la importancia de medir el desempeño de los algoritmos antes de decidir cuál utilizar en un proyecto de análisis de datos.

Para el proyecto, esta medición permite justificar que los algoritmos manuales son útiles para comprender la lógica de ordenamiento, pero que en un flujo analítico real conviene priorizar funciones optimizadas y documentadas cuando se trabaja con volúmenes mayores de información.

In [25]:
# Algoritmo recursivo: merge sort

def merge_sort(lista):
    """
    Ordena una lista utilizando el algoritmo recursivo merge sort.
    
    Caso base:
    Si la lista tiene 0 o 1 elemento, ya está ordenada.
    
    Caso recursivo:
    Se divide la lista en dos partes, se ordena cada parte
    y luego se combinan ambas listas ordenadas.
    """
    if len(lista) <= 1:
        return lista

    mitad = len(lista) // 2

    izquierda = merge_sort(lista[:mitad])
    derecha = merge_sort(lista[mitad:])

    return combinar_listas(izquierda, derecha)


def combinar_listas(izquierda, derecha):
    """
    Combina dos listas ordenadas en una sola lista ordenada.
    """
    resultado = []
    i = 0
    j = 0

    while i < len(izquierda) and j < len(derecha):
        if izquierda[i] <= derecha[j]:
            resultado.append(izquierda[i])
            i += 1
        else:
            resultado.append(derecha[j])
            j += 1

    resultado.extend(izquierda[i:])
    resultado.extend(derecha[j:])

    return resultado

In [26]:
# Prueba del algoritmo recursivo con datos reales del dataset

datos_merge = df["salary_after_ai"].head(20).tolist()

resultado_merge = merge_sort(datos_merge.copy())

print("Primeros 20 salarios originales:")
print(datos_merge)

print("\nPrimeros 20 salarios ordenados con merge_sort:")
print(resultado_merge)

Primeros 20 salarios originales:
[95455, 72013, 42290, 107820, 40945, 104781, 86183, 105776, 104197, 42699, 92231, 109956, 85267, 64752, 102088, 36482, 46193, 42537, 79219, 101637]

Primeros 20 salarios ordenados con merge_sort:
[36482, 40945, 42290, 42537, 42699, 46193, 64752, 72013, 79219, 85267, 86183, 92231, 95455, 101637, 102088, 104197, 104781, 105776, 107820, 109956]


In [27]:
# Medición de tiempo del algoritmo recursivo merge_sort

datos_prueba_merge = df["salary_after_ai"].head(500).tolist()

tiempo_merge = timeit.timeit(
    lambda: merge_sort(datos_prueba_merge.copy()),
    number=10
)

print("Tiempo de ejecución merge_sort:", tiempo_merge)
print("Cantidad de datos evaluados:", len(datos_prueba_merge))

Tiempo de ejecución merge_sort: 0.006144099999801256
Cantidad de datos evaluados: 500


In [28]:
# Comparación general de tiempos de ejecución

print("Comparación de tiempos sobre 500 registros de salary_after_ai")
print("----------------------------------------------------------")
print("Insertion sort:", tiempo_insertion)
print("Merge sort:", tiempo_merge)
print("Sorted Python:", tiempo_sorted)

print("\nRelación de tiempos:")
print("Insertion sort / Merge sort:", tiempo_insertion / tiempo_merge)
print("Merge sort / Sorted Python:", tiempo_merge / tiempo_sorted)
print("Insertion sort / Sorted Python:", tiempo_insertion / tiempo_sorted)

Comparación de tiempos sobre 500 registros de salary_after_ai
----------------------------------------------------------
Insertion sort: 0.0513978999988467
Merge sort: 0.006144099999801256
Sorted Python: 0.0002791999995679362

Relación de tiempos:
Insertion sort / Merge sort: 8.365407464154112
Merge sort / Sorted Python: 22.006088858557632
Insertion sort / Sorted Python: 184.08989999421664


## Comparación entre algoritmo estructurado, recursivo y función optimizada

En esta sección se compararon tres enfoques de ordenamiento aplicados sobre 500 registros reales de la variable `salary_after_ai`:

- `insertion_sort`: algoritmo estructurado implementado manualmente.
- `merge_sort`: algoritmo recursivo basado en la estrategia de dividir y vencer.
- `sorted()`: función optimizada incorporada en Python.

Los resultados muestran que `insertion_sort` fue el enfoque más lento, debido a que su complejidad temporal en el peor caso es O(n²). Esto significa que, al aumentar la cantidad de datos, el número de comparaciones puede crecer de forma considerable.

El algoritmo `merge_sort` obtuvo un mejor desempeño que `insertion_sort`, ya que utiliza recursividad para dividir la lista en partes más pequeñas, ordenar cada parte y luego combinarlas. Su complejidad temporal esperada es O(n log n), por lo que resulta más eficiente para conjuntos de datos más grandes.

Finalmente, `sorted()` fue la alternativa más rápida, ya que corresponde a una función interna de Python altamente optimizada. Esta comparación permite concluir que los algoritmos manuales son útiles para comprender la lógica de programación estructurada y recursiva, pero que en un proyecto de análisis de datos conviene utilizar funciones optimizadas cuando se requiere eficiencia y escalabilidad.

Esta medición responde al objetivo de la Fase 3, ya que permite fundamentar decisiones de diseño algorítmico a partir de evidencia de ejecución y no solo desde una apreciación teórica.

In [29]:
# Validación final del avance Fase 3

validacion_f3 = {
    "dataset_cargado": df is not None,
    "filas_dataset": df.shape[0],
    "columnas_dataset": df.shape[1],
    "algoritmo_estructurado": "insertion_sort ejecutado",
    "algoritmo_recursivo": "merge_sort ejecutado",
    "medicion_insertion_sort": tiempo_insertion,
    "medicion_merge_sort": tiempo_merge,
    "medicion_sorted": tiempo_sorted
}

validacion_f3

{'dataset_cargado': True,
 'filas_dataset': 2000,
 'columnas_dataset': 17,
 'algoritmo_estructurado': 'insertion_sort ejecutado',
 'algoritmo_recursivo': 'merge_sort ejecutado',
 'medicion_insertion_sort': 0.0513978999988467,
 'medicion_merge_sort': 0.006144099999801256,
 'medicion_sorted': 0.0002791999995679362}

In [30]:
# Exportación de resultados de complejidad a archivo TXT

F3_RESULTS.mkdir(parents=True, exist_ok=True)

with open(RESULTS_COMPLEXITY, "w", encoding="utf-8") as archivo:
    archivo.write("Mediciones de complejidad - Fase 3\n")
    archivo.write("=================================\n\n")
    archivo.write("Variable analizada: salary_after_ai\n")
    archivo.write("Cantidad de registros evaluados: 500\n\n")
    archivo.write(f"Tiempo insertion_sort: {tiempo_insertion}\n")
    archivo.write(f"Tiempo merge_sort: {tiempo_merge}\n")
    archivo.write(f"Tiempo sorted(): {tiempo_sorted}\n\n")
    archivo.write("Interpretación:\n")
    archivo.write("Insertion sort fue el algoritmo más lento debido a su complejidad O(n²).\n")
    archivo.write("Merge sort fue más eficiente, con complejidad esperada O(n log n).\n")
    archivo.write("sorted() fue la opción más rápida por estar optimizada internamente en Python.\n")

print("Archivo de resultados guardado en:", RESULTS_COMPLEXITY)
print("Archivo existe:", RESULTS_COMPLEXITY.exists())

Archivo de resultados guardado en: C:\Users\LuiskarEspinoza\Documents\Proyectos\mi_app\mcdi500_s1_grupo6\F3\results\mediciones_complejidad_f3.txt
Archivo existe: True


## Conclusión técnica del avance Fase 3

En este avance se implementaron y evaluaron algoritmos de programación estructurada y recursiva sobre el dataset proveniente de la fase 2, como se puede evidenciar en este notebook. Se crea y se utiliza la variable `salary_after_ai` como caso de estudio, se implementaron y compararon distintos enfoques de ordenamiento, utilizando datos reales del proyecto.

El algoritmo `insertion_sort` permitió representar una solución estructurada y comprensible, aunque con menor eficiencia para conjuntos de datos más grandes. Por su parte, `merge_sort` permitió incorporar recursividad mediante una estrategia de división del problema en subproblemas más pequeños, obteniendo un mejor desempeño que el algoritmo estructurado manual.

La comparación con `sorted()` permitió observar la importancia de utilizar funciones optimizadas cuando se requiere eficiencia computacional. Esta evidencia refuerza la idea discutida en el foro técnico de la Semana 1: las decisiones algorítmicas deben fundamentarse en el tipo de problema, el volumen de datos y las mediciones de desempeño, no solo en la preferencia por una técnica específica.

Como proyección, el código puede evolucionar hacia una organización más modular, trasladando las funciones principales al archivo `F3/src/algoritmos_f3.py` y documentando los resultados en el repositorio GitHub.